happi/cli.py

```diff
@@ -102,13 +105,58 @@ def happi_cli(args):
                 return
             if value.replace('.', '').isnumeric():
                 logger.debug('Changed %s to float', value)
-                value = float(value)
-            client_args[criteria] = value
+                value = str(float(value))
 
-        results = client.search(**client_args)
-        if results:
+            if criteria == 'z' and ',' in value:
+                start = None
+                end = None
+                z_is_range = True
+                try:
+                    start, end = value.replace(
+                        '[', '').replace(']', '').split(',')
+                    start = float(start)
+                    end = float(end)
+                except Exception as ex:
+                    logger.error("Invalid numbers for the z range %s", ex)
+                    sys.exit(1)
+                if start < end:
+                    results += client.search_range('z', start, end)
+                else:
+                    logger.error('Invalid z range')
+
+            # skip the criteria for range values
+            # it won't be a valid criteria for search_regex()
+            if z_is_range and criteria == 'z':
+                pass
+            else:
+                client_args[criteria] = fnmatch.translate(value)
+
+        results += client.search_regex(**client_args)
+
+        # find the repeated items
+        res_size = len(results)
+        repeated = []
+        for i in range(res_size):
+            k = i + 1
+            for j in range(k, res_size):
+                if results[i] == results[j] and results[i] not in repeated:
+                    repeated.append(results[i])
+
+        # we only want to return the ones that have been repeated
```

Comment:
This checking for repeats approach assumes that we are doing no more than two searches, so it will need to be refactored if we find ourselves doing more (for example, multiple range searches on different fields)

Reply:
e.g. we'll need to switch to keeping the results from every search separate, then doing the `and` of all the sets

In general, using sets like this would probably clean the combination logic quite a bit

Reply:
Yeah I think I need to think about this guy a bit more.... I initially tried to have them in separated lists, but I was kind of running into the same logic mess, because of the multiple scenarios:
1. when using both `search_range` + `search_regex`
    -  I need to return only the ones that they both matched
2. when using only `search_range` 
   - I need to know if i used the `search_regex` as well, if i check for an empty search_regex list is not enough, since i possible did not match anything 
3. when using only `search_regex`
    - here is the same situation as in number 2 - i need an extra variable to know if i have used `search_range` as well...

I also tried to use the sets idea too - something like: `set(a) & set(b)` to give me the ones that are equal in both lists, but I was getting an `unhashable type` error for `SearchResult`  - i though i was just not able to add a dictionary into the set, but I've seen people casting the item into a tuple first, i could give it a try, something like this: 
```
a = {1:2, 3:4}
s = set()
s.add(tuple(a.items()))
```

Maybe I should change my approach here completely... need to think about it - thanks a lot for this feedback 👍

Reply:
Interesting, I would have thought that a `SearchResult` would be hashable... That is unfortunate. That sounds like the cleanest way to me, create a list of sets and collapse them with &.

To be clear, I would merge this PR as-is since it is already very useful. Just adding food for thought, anything can be addressed now or we can make a new issue to address it later.

Reply:
Sounds good @ZLLentz  - I am trying now to make it a bit cleaner with the use of sets - i agree with you on this, it will look cleaner :)

Reply:
Still working on this one ... i am sorry it takes me so long to get this figured out, It might be that I don't know what i am doing here, I can't get those items in a `Set` like I wanted (casting into tuples) - and I am trying to come up with a cleaner way i can do that with but anything i try gets messy, my brain is not thinking well - please give me a bit more time on this and hopefully i'll come up with something a bit cleaner ?